power_grid_model_c/power_grid_model/include/power_grid_model/grouped_index_vector.hpp

```diff
@@ -0,0 +1,148 @@
+// SPDX-FileCopyrightText: 2022 Contributors to the Power Grid Model project <dynamic.grid.calculation@alliander.com>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+#ifndef POWER_GRID_MODEL_SPARSE_IDX_VECTOR_HPP
+#define POWER_GRID_MODEL_SPARSE_IDX_VECTOR_HPP
+
+#include "power_grid_model.hpp"
+
+#include <boost/iterator/iterator_facade.hpp>
+#include <boost/iterator/zip_iterator.hpp>
+#include <boost/range.hpp>
+#include <boost/range/counting_range.hpp>
+
+/*
+A data-structure for iterating through the indptr, ie. sparse representation of data.
+Indptr can be eg: [0, 3, 6, 7]
+This means that:
+objects 0, 1, 2 are coupled to index 0
+objects 3, 4, 5 are coupled to index 1
+objects 6 is coupled to index 2
+
+Another intuitive way to look at this for python developers is like list of lists: [[0, 1, 2], [3, 4, 5], [6]].
+
+DenseIdxVector is a vector of element to group. ie. [0, 1, 1, 4] would denote that [[0], [1, 2], [], [], [3]]
+The input, ie. [0, 1, 3] should be strictly increasing
+
+*/
+
+namespace power_grid_model {
+
+class SparseIdxVector {
+  private:
+    class GroupIterator : public boost::iterator_facade<GroupIterator, Idx, boost::random_access_traversal_tag,
+                                                        boost::iterator_range<IdxCount>, Idx> {
+      public:
+        explicit constexpr GroupIterator(IdxVector const& indptr, Idx group) : indptr_(indptr), group_(group) {}
```

Comment:
it's probably a good idea to have constructors with tags (similar to how it's done for the `Deserializer`). that way, the converter we wrote for the tests can be used to easily construct one from the other